### ISSUE TITLE
Cannot build application on HHVM

### ISSUE THREAD

**Issue by Danon** [sebastianbergmann/phpunit#3320]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | `^6.4`
| PHP version         | HipHop VM 3.28.3
| Installation Method | Composer

While running a travis build for `hhvm` this is an error I get
```bash
Fatal error: Uncaught Error: Parameter $constraints is variadic and has a type constraint (Constraint); variadic params with type constraints are not supported in non-Hack files in phar://phpunit-6.5.2.phar/phpunit/Framework/Constraint/LogicalAnd.php:29
Stack trace:
#0 /home/travis/.phpenv/versions/hhvm-stable/bin/phpunit(103): include()
#1 {main}
```

What can I do to run tests in PhpUnit on hhvm?

**Comment by Danon** [comment]
```bash
$ composer info | sort
doctrine/instantiator              1.1.0  A small, lightweight utility to...
guzzlehttp/guzzle                  6.3.3  Guzzle is a PHP HTTP client lib...
guzzlehttp/promises                v1.3.1 Guzzle promises library
guzzlehttp/psr7                    1.4.2  PSR-7 message implementation th...
myclabs/deep-copy                  1.8.1  Create deep copies (clones) of ...
phar-io/manifest                   1.0.1  Component for reading phar.io m...
phar-io/version                    1.0.1  Library for handling version in...
phpdocumentor/reflection-common    1.0.1  Common reflection classes used ...
phpdocumentor/reflection-docblock  4.3.0  With this component, a library ...
phpdocumentor/type-resolver        0.4.0 
phpspec/prophecy                   1.8.0  Highly opinionated mocking fram...
phpunit/php-code-coverage          5.3.2  Library that provides collectio...
phpunit/php-file-iterator          1.4.5  FilterIterator implementation t...
phpunit/php-text-template          1.2.1  Simple template engine.
phpunit/php-timer                  1.0.9  Utility class for timing
phpunit/php-token-stream           2.0.2  Wrapper around PHP's tokenizer ...
phpunit/phpunit                    6.5.13 The PHP Unit Testing framework.
phpunit/phpunit-mock-objects       5.0.10 Mock Object library for PHPUnit
psr/http-message                   1.0.1  Common interface for HTTP messages
psr/log                            1.0.2  Common interface for logging li...
satooshi/php-coveralls             v2.0.0 PHP client library for Coverall...
sebastian/code-unit-reverse-lookup 1.0.1  Looks up which function or meth...
sebastian/comparator               2.1.3  Provides the functionality to c...
sebastian/diff                     2.0.1  Diff implementation
sebastian/environment              3.1.0  Provides functionality to handl...
sebastian/exporter                 3.1.0  Provides the functionality to e...
sebastian/global-state             2.0.0  Snapshotting of global state
sebastian/object-enumerator        3.0.3  Traverses array structures and ...
sebastian/object-reflector         1.1.1  Allows reflection of object att...
sebastian/recursion-context        3.0.0  Provides functionality to recur...
sebastian/resource-operations      1.0.0  Provides a list of PHP built-in...
sebastian/version                  2.0.1  Library that helps with managin...
symfony/config                     v4.1.5 Symfony Config Component
symfony/console                    v4.1.5 Symfony Console Component
symfony/filesystem                 v4.1.5 Symfony Filesystem Component
symfony/polyfill-ctype             v1.9.0 Symfony polyfill for ctype func...
symfony/polyfill-mbstring          v1.9.0 Symfony polyfill for the Mbstri...
symfony/stopwatch                  v4.1.5 Symfony Stopwatch Component
symfony/yaml                       v4.1.5 Symfony Yaml Component
theseer/tokenizer                  1.1.0  A small library for converting ...
webmozart/assert                   1.3.0  Assertions to validate method i...
```

**Comment by sebastianbergmann** [comment]
HHVM is not supported.

**Comment by Danon** [comment]
Yes, I found that out, but is there anything I can do to run the tests on HHVM? Can I downgrade it to some version or something?

Is there no way to run it, other than to rewrite my tests to other testing framework?

**Comment by ayg** [comment]
HHVM has announced that [they no longer intend to maintain compatibility with PHP](https://hhvm.com/blog/2017/09/18/the-future-of-hhvm.html), so if you want to stay on HHVM, you should not expect to be able to use anything that's written for PHP.  You should commit your app to either PHP 7 without HHVM compatibility, or pure Hack without PHP compatibility.  Trying to work with both is going to cause a lot of headaches for you, and will eventually be impossible.

**Comment by Danon** [comment]
@ayg Yes, I'm aware of that. However, my applications still work under HHVM, so I see no reason to stop building it on HHVM.

The only thing, that's stopping me is PhpUnit, which can't be run on hhvm. If I were to rewrite all my tests to other Php testing frameworks, everything would work as expected

**Comment by Danon** [comment]
@epdenouden I wanted to do it! Really! but:
- I can't find and older version of PhpUnit, they all use variadic arguments which aren't compatible with HHVM
- I couldn't find a single test framework for HHVM, maybe you know one?
- Isn't PhpUnit 7.4 supposed to work on PHPv7? Cause it still doesn't work with hhvm.

Sorry, but the project I'm maintaining has to work for Php v7 and it would be nice if it worked on HHVM and **it does**! But phpunit doesn't.

**Comment by epdenouden** [comment]
Hello @Danon!

Yes, PHPUnit in fact requires PHP 7.1 or higher. Because we want to use every bit of performance we can get and the new language features are nice, too. Since PHP7 is the way forward HHVM support has actually *been removed* from PHPUnit a while ago.

I do not know of any other frameworks that keep support for either PHP5 or HHVM alive. Keep in mind that both those technologies are almost at their end-of-life.
- PHP5 and PHP7 end of 2018: http://php.net/supported-versions.php
- HHVM for PHP5: https://hhvm.com/blog/2017/09/18/the-future-of-hhvm.html

I am wondering, is it really that much work to update your tests to work with PHPUnit 7.x?

**Comment by epdenouden** [comment]
@Danon The short answer is: _you cannot run your PHPUnit tests on the HHVM_

I understand you want your project to run both on PHP v7 and the HHVM. The closest you can come to a 'solution' would be: write extensive tests for your library (the regular expression one?) and make sure it's 100% coverage and green, so you are pretty sure everything works as expected.

Then cover the HHVM runtime context with a list of integration tests along the lines of: `yourRegExTest($input) !== $expectedOutput`.

That way you make sure
- the business logic and flow of your project has maximum coverage under normal PHP
- when running the HHVM the output at least is the same as the PHP-only version

I hope this helps.

**Comment by epdenouden** [comment]
BTW, keep in mind that this will complicate your CI: now you will have to run your project through two testing unit test passes, each with a slightly different PHP version+config.

**Comment by epdenouden** [comment]
Ok, I will rephrase and answer your exact questions:
- I personally do not know what the last version of PHPUnit was that supports the HHVM. To find out, search for the phrase 'HHVM' in the README files of previous versions. Myself I only started contributing starting with PHPUnit v7 and as far as I know HHVM is looooong ago.
- I never researched HHVM compatibility of other testing frameworks. However, there aren't many such frameworks for PHP in use today, so Google is your friend here.
- If your tests are running on PHPUnit 7.4 already, you could write a quick'n'dirty integration tests for the HHVM very easily! Just capture the input and actual values for the asserts and you have a basic set. Taa-daa :)

For more context, search for _HHVM_ in the issues here on Github. For example: https://github.com/sebastianbergmann/phpunit/issues/3207

**Comment by siraic** [comment]
- HHVM never aimed to support PHP7, only PHP5. Any support for PHP 7 code is purely coincidental. They are now on a path to remove PHP support altogether and fully focus on Hack in future versions.
- PHPUnit 6 and up only supports PHP7, PHPUnit 5 supported both 5.6 and 7, but since typed variadic function arguments (introduced in 5.6) were never supported in HHVM, I'm not sure that version is going to work. PHPUnit 4 supported PHP5.5 (so no variadic arguments can be used), but officially didn't support PHP7, though you might find a version working with it.

If you are set on keeping support for HHVM, I suggest you branch off from your master branch now, lock your PHP version to 5.5|5.6 in your composer file, use (unsupported) PHPUnit 4 and keep supporting PHP 5.5 and HHVM in this separate version of your code. Maybe right now only PHPUnit is preventing HHVM compatibility (or it might just be the first fatal error to pop up, while there are way more incompatibilities hidden behind it), but many more libraries are going be unsupported soon enough. It is not going to be worth the headaches trying to juggle dependencies that accidentally supported HHVM as well as recent PHP versions.

**Comment by clue** [comment]
> @Danon So I have questions:
> 
> * What version of phpunit keeps working with hhvm?
> 
> * Please, tell me what testing framework is fine with hhvm?
> 
> * My tests are already phpunit 7.4 and are running on php7.1 and php7.2 - what now?

While I agree with the above objections that HHVM decided to drop PHP support in the future and I do not expect PHPUnit to keep supporting HHVM, just to answer your question, here's how you can run an outdated and unsupported PHPUnit version with HHVM: Outdated PHPUnit 5 appears to be the last version that I could get to run on outdated HHVM on Travis CI. You can also explicitly install another PHPUnit version just for HHVM using Travis CI, for example see also https://github.com/clue/php-socket-raw/pull/42.

**Comment by Danon** [comment]
So there is no testing framework for PHP for HHVM?

**Comment by filips123** [comment]
What is the latest version that supported HHVM? Does it support namespaced test syntax (available after PHPUnit 4.8.36)?
